File: python/101-150/109. Convert Sorted List to Binary Search Tree.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        if not head:
            return None
        temp = []
        cur = head
        while cur:
            temp.append(cur.val)
            cur = cur.next
        return self.helper(0, len(temp) - 1, temp)
    def helper(self, low, hi, sortedList):
        if low > hi:
            return None
        mid = (hi - low)/2  + low
        root = TreeNode(sortedList[mid])
        root.left = self.helper(low, mid - 1, sortedList)
        root.right = self.helper(mid + 1, hi, sortedList)
        return root
        

File: cpp/101-150/109. Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	TreeNode* sortedListToBST(ListNode* head) {
		return helper(head, NULL);
	}
	TreeNode* helper(ListNode *begin, ListNode *end){
		if(begin == end)
			return NULL;
		if(begin->next == end){
			TreeNode *root = new TreeNode(begin->val);
			return root;
		}
		ListNode *slow, *fast;
		slow = fast = begin;
		while(fast != end && fast->next != end){
			slow = slow->next;
			fast = fast->next->next;
		}
		ListNode *mid = slow;
		TreeNode *root = new TreeNode(mid->val);
		root->left = helper(begin, mid);
		root->right = helper(mid->next, end);
		return root;
	}
};